File: src/main/java/com/hp/jipp/model/PrinterState.java
package com.hp.jipp.model;

import com.google.auto.value.AutoValue;
import com.hp.jipp.encoding.NameCode;
import com.hp.jipp.encoding.NameCodeType;

/** An enumeration of possible printer states */
@AutoValue
public abstract class PrinterState extends NameCode {

    public static final PrinterState Idle = of("idle", 3);
    public static final PrinterState Processing = of("processing", 4);
    public static final PrinterState Stopped = of("stopped", 5);

    public static final NameCodeType.Encoder<PrinterState> ENCODER = NameCodeType.Encoder.Companion.of(
            PrinterState.class, new NameCode.Factory<PrinterState>() {
                @Override
                public PrinterState of(String name, int code) {
                    return PrinterState.of(name, code);
                }
            });

    public static PrinterState of(String name, int code) {
        return new AutoValue_PrinterState(name, code);
    }
}

File: src/main/java/com/hp/jipp/model/PrinterState.kt
// Copyright 2017 HP Development Company, L.P.
// SPDX-License-Identifier: MIT

package com.hp.jipp.model

import com.hp.jipp.encoding.Enum
import com.hp.jipp.encoding.EnumType

/** An enumeration of possible printer states  */
data class PrinterState(override val code: Int, override val name: String) : Enum() {

    /** An attribute type for [PrinterState] objects */
    class Type(name: String) : EnumType<PrinterState>(Encoder, name)

    companion object {
        @JvmField val idle = PrinterState(3, "idle")
        @JvmField val processing = PrinterState(4, "processing")
        @JvmField val stopped = PrinterState(5, "stopped")

        @JvmField val Encoder = EnumType.Encoder(PrinterState::class.java) { code, name ->
            PrinterState(code, name)
        }
    }
}
